### PR TITLE
Refactoring/Defensive-programming

### PR DESCRIPTION
**Pull request by alexcrt**
- Moved selectors and function to their respective package (see issue #14)
- Added checks to constructor parameters for the spliterators 
- Make all the spliterator classes final

### CODE REVIEW

**Comment by alexcrt** [comment]
@poetix When you have time, could you take a look at the Travis build?

File: src/protonpack/java/com/codepoetics/protonpack/AggregatingSpliterator.java [hunk 1]
```diff
@@ -15,8 +17,8 @@
     private List<I> currentSlide = new ArrayList<>();
 
     AggregatingSpliterator(Spliterator<I> source, BiPredicate<List<I>, I> predicate) {
-        this.source = source;
-        this.condition = predicate;
+        this.source = requireNonNull(source);
+        this.condition = requireNonNull(predicate);
```
Comment (poetix) [inline src/protonpack/java/com/codepoetics/protonpack/AggregatingSpliterator.java]: I like to follow the convention that constructors do no work (besides assignment), and cannot throw exceptions. So I would ordinary create a static factory method to do work like validation in.

**Comment by poetix** [comment]
You seem somehow to have renamed src/main/java to src/protonpack/java, so it's not surprising that travis isn't building...

File: src/protonpack/java/com/codepoetics/protonpack/AggregatingSpliterator.java (hunk 1 above)
Comment (alexcrt) [inline src/protonpack/java/com/codepoetics/protonpack/AggregatingSpliterator.java]: Hmm ok, we can still move it into the factory method but if they are many that could lead to code duplication... Well feel free to reject this PR :-)

**Comment by alexcrt** [comment]
Holy.... I didn't pay attention to that. Intellij probably renamed it and I didn't see....

**Comment by poetix** [comment]
Yes, it's a good idea to do a command-line mvn clean install before committing, just to be sure.

File: src/protonpack/java/com/codepoetics/protonpack/AggregatingSpliterator.java (hunk 1 above)
Comment (poetix) [inline src/protonpack/java/com/codepoetics/protonpack/AggregatingSpliterator.java]: I think I will in this case. I'm not convinced of the value of null checks. Far better not to go passing nulls around in the first place. If you pass null as a parameter into something, my view is that you deserve everything you get. A standardised @NotNull annotation, compiler-checked, would be worth having; but my preference would be for non-null to be standard, and nullibility to be clearly marked as an exception.

Comment (alexcrt) [inline src/protonpack/java/com/codepoetics/protonpack/AggregatingSpliterator.java]: There's still the windowed spliterator. If you give some dumb parameters, such as a negative window size, it hangs in an infinite loop. I don't think it's a desirable behavior.

**Pull request closed without merging**